src/modules/artnet/consumer/artnet_consumer.cpp

```diff
@@ -0,0 +1,298 @@
+
+                    compute_fixtures();
+                }
+
+
+                void initialize(const core::video_format_desc& /*format_desc*/, int /*channel_index*/) override {
+                    thread_ = std::thread([this] {
+                        try {
+                            long long time = 1000 / config.refreshRate;
+                            auto last_send = std::chrono::system_clock::now();
+
+                            while (!abort_request_) {
+                                auto now = std::chrono::system_clock::now();
+                                std::chrono::duration<double> elapsed_seconds = now - last_send;
+                                long long elapsed_ms = std::chrono::duration_cast<std::chrono::milliseconds>(elapsed_seconds).count();
+
+                                long long sleep_time = time - elapsed_ms * 1000;
+                                if (sleep_time > 0) std::this_thread::sleep_for(std::chrono::milliseconds(sleep_time));
+
+                                last_send = now;
+
+                                frame_mutex_.lock();
+                                auto frame = last_frame_;
+
+                                frame_mutex_.unlock();
+                                if (!frame) continue; // No frame available
+
+                                uint8_t dmx_data[512];
+                                memset(dmx_data, 0, 512);
+
+                                for (auto computed_fixture : computed_fixtures) {
+                                    auto color = average_color(frame, computed_fixture.rectangle);
+                                    uint8_t* ptr = dmx_data + computed_fixture.address;
+
+                                    switch (computed_fixture.type) {
+                                        case FixtureType::DIMMER:
+                                            ptr[0] = (uint8_t) (0.279 * color.r + 0.547 * color.g + 0.106 * color.b);
+                                            break;
+                                        case FixtureType::RGB:
+                                            ptr[0] = color.r;
+                                            ptr[1] = color.g;
+                                            ptr[2] = color.b;
+                                            break;
+                                        case FixtureType::RGBW:
+                                            uint8_t w = std::min(std::min(color.r, color.g), color.b);
+                                            ptr[0] = color.r - w;
+                                            ptr[1] = color.g - w;
+                                            ptr[2] = color.b - w;
+                                            ptr[3] = w;
+                                            break;
+                                    }
+                                }
+
+                                send_dmx_data(dmx_data, 512);
+                            }
+                        } catch (...) {
```

Comment:
should this be outside or inside of the while loop?

Currently, when something throws it will be caught outside of the while loop, causing the thread to terminate. The consumer will then be left not outputting any data, appearing to casparcg, AMCP and OSC to be still running.

This be moved inside of the while loop, so that it will continue to try to send data on the next frame. Or alternatively, if you store and rethrow the error from `send()`, then we will see that the consumer errored and will automatically remove it. (I think, I am working from memory of the decklink consumer)